Kuroba/app/src/main/java/com/github/adamantcheese/chan/features/gesture_editor/EditableZone.kt

```diff
@@ -0,0 +1,267 @@
+     * */
+    fun onTouchInProgress(x: Float, y: Float, dx: Float, dy: Float): Boolean {
+        if (isMoving) {
+            // We are moving the whole zone
+            val newX = x - zoneOriginToTouchPosDeltaWhenMoving.x
+            val newY = y - zoneOriginToTouchPosDeltaWhenMoving.y
+
+            onMoving(newY, newX)
+            updateHandlePosition()
+
+            return true
+        }
+
+        if (isResizing) {
+            // We are resizing the zone by moving the handle
+            onResizing(dx, dy)
+            updateHandlePosition()
+
+            return true
+        }
+
+        return false
+    }
+
+    fun onTouchEnd() {
+        isMoving = false
+        isResizing = false
+    }
+
+    fun getCurrentAttachSide(): AttachSide = currentAttachSide
+    fun getCurrentZone(): SafeRectF = zone
+
+    private fun onMoving(newY: Float, newX: Float) {
+        when (currentAttachSide) {
+            AttachSide.Left,
+            AttachSide.Right -> {
+                zone.moveTo(zone.x, newY)
+
+                if (zone.top() < 0f) {
+                    zone.moveTo(zone.x, 0f)
+                }
+                if (zone.bottom() > viewHeight) {
+                    zone.moveTo(zone.x, viewHeight - zone.height)
+                }
+            }
+            AttachSide.Top,
+            AttachSide.Bottom -> {
+                zone.moveTo(newX, zone.y)
+
+                if (zone.left() < 0f) {
+                    zone.moveTo(0f, zone.y)
+                }
+                if (zone.right() > viewWidth) {
+                    zone.moveTo(viewWidth - zone.width, zone.y)
+                }
+            }
+        }
+    }
+
+    private fun onResizing(dx: Float, dy: Float) {
+        when (currentAttachSide) {
+            AttachSide.Left -> {
+                val targetHeight = zone.height + dy
+                val realNewHeight = targetHeight.coerceIn(MINIMUM_SIZE, MAXIMUM_SIZE)
+
+                zone.setHeight(realNewHeight)
+            }
+            AttachSide.Right -> {
+                val targetHeight = zone.height + dy
+                val realNewHeight = targetHeight.coerceIn(MINIMUM_SIZE, MAXIMUM_SIZE)
+
+                zone.setHeight(realNewHeight)
+            }
+            AttachSide.Top -> {
+                val targetWidth = zone.width - dx
+                val realNewWidth = targetWidth.coerceIn(MINIMUM_SIZE, MAXIMUM_SIZE)
+
+                zone.moveTo((zone.x + dx) - (realNewWidth - targetWidth), zone.y)
+                zone.setWidth(realNewWidth)
+            }
+            AttachSide.Bottom -> {
+                val targetWidth = zone.width - dx
+                val realNewWidth = targetWidth.coerceIn(MINIMUM_SIZE, MAXIMUM_SIZE)
+
+                zone.moveTo((zone.x + dx) - (realNewWidth - targetWidth), zone.y)
+                zone.setWidth(realNewWidth)
+            }
+        }
+    }
+
+    private fun updateHandlePosition() {
```

Comment:
This function and the one below it are essentially identical, probably a good idea to reuse the code if possible.

Reply:
Done.